src/controls/framework/defineControl.ts

```diff
@@ -0,0 +1,159 @@
+import { IControlDescriptor } from './controlMetadata';
+import { Control } from '../Control';
+
+type LocalizedString = string;
+
+export enum PropertyType {
+  unknown,
+  string = 1 << 1,
+  number = 1 << 2,
+  enum = 1 << 3,
+  boolean = 1 << 4,
+}
+
+interface IPropertyInfo {
+  id: string;
+  displayName: LocalizedString;
+  propertyType: PropertyType;
+}
+
+interface IGetSetProperty<TOrigin, T> {
+  getValue(state: TOrigin): T;
+  setValue(state: TOrigin, value: T);
+}
+
+export interface IOwnedProperty<TState, T> extends IPropertyInfo, IGetSetProperty<TState, T> {
+  getEditor?: (instance: ControlContainer) => IPropertyEditor;
+}
+
+type Transformer<TFrom, TTo> = (state: TFrom) => TTo;
+
+/**
+ * Creates a property of the form IOwnedProperty<TState, TTo> by wrapping an existing property that operates on TFrom
+ * to TTo by using a callback to transform TState into TFrom.
+ */
+class WrappedProperty<TState, TFrom, TTo> implements IOwnedProperty<TState, TTo> {
+  constructor(private getter: Transformer<TState, TFrom>, private property: IOwnedProperty<TFrom, TTo>) {}
+
+  get displayName(): LocalizedString {
+    return this.property.displayName;
+  }
+
+  get id(): string {
+    return this.property.id;
+  }
+
+  get propertyType(): PropertyType {
+    return this.property.propertyType;
+  }
+
+  getValue(state: TState): TTo {
+    return this.property.getValue(this.getter(state));
+  }
+
+  setValue(state: TState, value: TTo) {
+    return this.property.setValue(this.getter(state), value);
+  }
+}
+
+export function createControlDefinition<TControl extends IControl, TState>(data: {
+  id: string;
+  displayName: LocalizedString;
+  factory: () => TState;
+}): IStateCreator<TControl, TState> {
+  return new NewControlDescriptor(data.id, data.displayName, data.factory);
+}
+
+class NewControlDescriptor implements IStateCreator<any, any> {
```

Comment:
This was ultimately the framework way to define a new control; the code is mostly readable - the reason for the functions like `createControlDefinition` and many of the other interfaces in the file was to get typescript to aid in defining the control in a fluent-way.

Ultimately, however, this was one of the reasons we ditched it - it made everything too complex.  Currently controls have a boilerplate, but that boilerplate is much more understandable